### PR TITLE
Fix issue in API `POST /services/(id or name)/update`

### LINKED ISSUE #26090: Can't update service through api /services/name/update
**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 7
 Running: 4
 Paused: 0
 Stopped: 3
Images: 13
Server Version: 1.12.0
Storage Driver: devicemapper
 Pool Name: docker-8:5-59899908-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 2.571 GB
 Data Space Total: 107.4 GB
 Data Space Available: 104.8 GB
 Metadata Space Used: 5.857 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.142 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /data/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /data/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: active
 NodeID: d29p5fdm4kj8frq85bdc14pad
 Is Manager: true
 ClusterID: b44p2m6k3592p8d9p3qbg9xeo
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot interval: 10000
  Heartbeat tick: 1
  Election tick: 3
 Dispatcher:
  Heartbeat period: 5 seconds
 CA configuration:
  Expiry duration: 3 months
 Node Address: 172.16.10.19
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.28.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 30.96 GiB
Name: localhost.localdomain
ID: 4Y6T:NCRV:K7J7:WOKO:ZIZY:LXYQ:IQRL:FMNL:PH4E:BPBX:6VU6:Y3ZB
Docker Root Dir: /data/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 65
 Goroutines: 178
 System Time: 2016-08-29T10:17:34.056627271+08:00
 EventsListeners: 4
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 docker.aispeech.com
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

physical

**Steps to reproduce the issue:**
1.POST /services/service name/update
2.POST /services/service id/update

**Describe the results you received:**
IN API v1.24
when POST /services/service name/update,return:
{"message":"rpc error: code = 5 desc = service ""service name"" not found"}
when POST /services/service id/update, it works

Is this a BUG?

### PR DESCRIPTION
**Pull request by yongtang**
**\- What I did**

This fix tries to address the issue raised in #26090 where remote API `POST /services/(id or name)/update` cannot use `name` to update. This is not consistent with the documentation of the remote API.

**\- How I did it**

This fix fixes this issue by performing a lookup with `getService` in case `name` instead of `id` is used in API.

**\- How to verify it**

This fix adds an integration test to cover the changes.

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #26090.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

### CODE REVIEW

File: daemon/cluster/cluster.go [hunk 1]
```diff
@@ -913,6 +913,12 @@ func (c *Cluster) UpdateService(serviceID string, version uint64, spec types.Ser
 		return err
 	}
 
+	// serviceID could be either ID or name so lookup is needed
+	currentService, err := getService(ctx, c.client, serviceID)
```
Comment (tonistiigi) [inline daemon/cluster/cluster.go]: Could you rename the `serviceID` variable instead of comment.

**Comment by tonistiigi** [comment]
LGTM

File: daemon/cluster/cluster.go (hunk 1 above)
Comment (yongtang) [inline daemon/cluster/cluster.go]: Thanks @tonistiigi. The pull request has been updated. Please let me know if there are any other issues.

File: ubuntu.log [hunk 2]
```diff
@@ -0,0 +1,26 @@
+Using default tag: latest
```
Comment (tonistiigi) [inline ubuntu.log]: @yongtang extra file

Comment (yongtang) [inline ubuntu.log]: @tonistiigi Oh I am so sorry about that! Did a `git add -A` and didn't pay enough attention. The PR has been updated now. Thanks for the help!

**Comment by vdemeester** [comment]
LGTM 🐮

**Comment by stevvooe** [comment]
This was intentional. If you don't have the service `ID`, you cannot update the service.

This should have been a fix to the documentation.

**Comment by tonistiigi** [comment]
ping @vieux

**Comment by yongtang** [comment]
@stevvooe @tonistiigi Sorry about this PR. Didn't know `ID` was intentional. I created another PR #26169 to revert this PR and updates the docs instead. Feel free to close #26169 if it does not match the intended behavior.

**Comment by ziqingzhu** [comment]
@stevvooe but we can also get the service ID by service name through Inspect GET /services/(id or name)

**Comment by stevvooe** [comment]
@yongtang @ziqingzhu The main issue is that it requires names to be unique and limits the ability to move to a namespace-based model in the future. If we require strict ID for updates, it makes this slightly more flexible, although there are API use cases that can be made easier if you use names.

I am not sure if #26169 is the right model or not, but this shouldn't be a random decision.

**Comment by tonistiigi** [comment]
@stevvooe We need to change service inspect and remove also then.

**Comment by stevvooe** [comment]
@tonistiigi Given that, would it be safe to say that we need to follow the `id|name` model, across the board? In general, that is probably okay, but we need to understand the implications:

1. Re-naming is not possible.
2. IDs are less useful since names fulfill their role (maybe we can lose complexity and axe ids completely).
3. Namespaces will have to be some magic syntax (`<namespace>.<name>`).

There are probably some others but these are minor.

**Comment by tonistiigi** [comment]
@stevvooe I think renaming is something we shouldn't do even if we can. Namespaces may be valid use case - but it depends on how we want to implement them.

We could do this all on cli side by running a filter first to find the ID. I think that is much better API design but very inconsistent with all other similar endpoints we have for other objects. Also we have already shipped the name based api for inspect and remove in `v1.12`.

**Comment by stevvooe** [comment]
@tonistiigi Alright, let's leave it as is, then.

**Pull request merged**